doc/man/man1/jassc.1
.\" jassc man page
.TH JASSC 1 "APRIL 2014" jassc
.SH NAME
jassc \- JASS compiler
.SH SYNOPSIS
.B jassc [-fsyntax-only]
.I output file
.I input files
.B ...
.SH DESCRIPTION
.B jassc
is a free JASS compiler.

ChangeLog
2012-03-22 Tamino Dauth <tamino@cdauth.eu>
* editor: Class MpqPriorityList handles team colors and textures more safely now.
* editor: Dropped redundant storage of sources in class MpqPriorityList (ordered by their priority).
* editor: Improved class SourcesDialog.
* editor: Fixed entry resolution of class MpqPriorityList.
* mpq: Added initial workaround for missing BZh header.
* mpq: Some minor revisions and optimizations.
* mpq: std::logical_error is thrown when file key should be calculated on empty path.

2012-03-15 Tamino Dauth <tamino@cdauth.eu>
* Improved API documentation.
* map: Many formats do use Boost Pointer containers now.
* map: Fixed reading and writing of class Tree.
* map: Class file format implements default version and id reading and writing now and stores the version as member.
* map: Dropped unnecessary W3m* members.
* map: No exception is thrown anymore on not having the latest format version.

2012-03-08 Tamino Dauth <tamino@cdauth.eu>
* editor: Fixed module action checking.
* editor: Dropped KXmlGuiClient inheritance for class Module.
* editor: Added class ModuleToolBar which behaves similar to ModuleMenu.
* editor: Class Editor doesn't delete resources anymore (take care of them).
* editor: MpqPriorityList uses boost::ptr_map for team color and glow textures now.

2012-02-25 Tamino Dauth <tamino@cdauth.eu>
* editor: Fixed early rendering of ModelView.
* editor: ModelView uses x and y values on render window creation now.

2012-02-24 Tamino Dauth <tamino@cdauth.eu>
* mdlx: Finished revision.
* editor: Class OgreMdlx doesn't use member function mdlxId() anymore which was quite slow.

2012-02-23 Tamino Dauth <tamino@cdauth.eu>
* core: BaseVertex inherits Format and boost::array now.
* map: BasePosition inherits boost::array now.
* mdlx: Huge revision implementing new BasePosition template.

2012-02-22 Tamino Dauth <tamino@cdauth.eu>
* Added general platform.hpp file and moved some typedefs into it.
* Split up include statements again and removed most from utilities.hpp.
* Added headers color.hpp and vertex.hpp.
* mpq: Returns shared pointers in many functions now.
* mpq: Class Hash doesn't hold any file pointer anymore.
* map: Revised class Shadow.
* map: Versions use uint32 now.
* editor: Texture editor doesn't load OGRE texture anymore.
* editor: Replaced boost smart pointers by Qt smart pointers.
* blp: MIP maps are stored in pointer vector now.

2012-02-20 Tamino Dauth <tamino@cdauth.eu>
* blp: Added custom buffer size for writing JPEG data.
* blp: Considering free_in_buffer now when writing JPEG data.
* blp: Added default parameter constants and parameter "threads" to write and read functions.
* editor: Considers new default values and parameter "threads" now.
* editor: Added class TextureEditor::SaveDialog.
* mdlx: Fixed constructor calls of GroupdMdxBlock.
* editor: Fixed UI files header generation.

2012-02-19 Tamino Dauth <tamino@cdauth.eu>
* blp: Fixed marker size type.
* blp: Missing markers will lead to exceptions now.
* blp: Only writes tables for the first MIP map.
* blp: Required scanlines got image height that all scanlines are being written and read at once.

2012-02-18 Tamino Dauth <tamino@cdauth.eu>
* blp: Uses Boost pointer containers now.

2012-02-17 Tamino Dauth <tamino@cdauth.eu>
* Dropped transitive linking (which is default in CMake).

2012-02-12 Tamino Dauth <tamino@cdauth.eu>
* editor: Fixed enabling color palette and alpha channel actions in texture editor.
* editor: Fixed detection of save options for class Texture.

2012-02-06 Tamino Dauth <tamino@cdauth.eu>
* map: Fixed several bugs in reading process of class CustomUnits.

2012-02-04 Tamino Dauth <tamino@cdauth.eu>
* map: Separated class Value and added visitor classes.

2012-02-01 Tamino Dauth <tamino@cdauth.eu>
* editor: MpqPriorityList doesn't handle resource deletion anymore.
* map: Fixed modifications allocation of class CustomUnits.
* map: Improved Value usage.

2012-01-21 Tamino Dauth <tamino@cdauth.eu>
* editor: Editor::root() doesn't throw an exception on no render systems anymore.
* editor: Added texture editor MIME workaround.

2012-01-15 Tamino Dauth <tamino@cdauth.eu>
* map: Many formats are using std::vector now.
* editor: Revised class Resource.

2011-12-22 Tamino Dauth <tamino@cdauth.eu>
* Option MAP depends on BLP as well now.
* Replaced preprocessor errors by warnings which might make it easier to port wc3lib.
* editor: Fixed MpqPriorityList::tr() calls.

2011-11-25 Tamino Dauth <tamino@cdauth.eu>
* map: Improved parsing support in class Txt.

2011-10-28 Tamino Dauth <tamino@cdauth.eu>
* editor: Added class VariablesDialog.
* editor: Added class VariableDialog.
* Improved script "generateui.sh"
* editor: Initial variables support for trigger editor.
* map: Added class Txt.
* editor: MpqPriorityList supports reading *.txt files now by using KDE SC's KConfig class.
* map: Improved various classes.

2011-10-23 Tamino Dauth <tamino@cdauth.eu>
* map: Added class TriggerData.
* map: Revised many trigger related class to add trigger function parameter read support.

2011-10-22 Tamino Dauth <tamino@cdauth.eu>
* editor: Improved trigger editor.
* editor: Added class WindowsMenu and improved editor's actions and map handling.
* editor: Fixed bug in MpqPriorityList::removeSource().

2011-10-21 Tamino Dauth <tamino@cdauth.eu>
* editor: Improved trigger editor.

2011-10-20 Tamino Dauth <tamino@cdauth.eu>
* editor: Fixed UI layouts for modules.
* editor: Improved trigger editor.
* map: Improved trigger classes.
* plugins: MIP maps aren't generated anymore when BLP is loaded from QImage.

2011-10-19 Tamino Dauth <tamino@cdauth.eu>
* editor: Improved module and its API documentation.
* editor: Fixed module segmentation fault.
* editor: Fixed module window flag.
* plugins: Improved class BlpIOHandler.
* blp: blp::Blp::generateMipMaps() considers parameter "regenerate" now.
* blp: blp::Blp::MipMap uses boost::multi_array for color storage now which should improve its performance.

src/app/jassc.cpp
/***************************************************************************
 *   Copyright (C) 2010 by Tamino Dauth                                    *
 *   tamino@cdauth.eu                                                      *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include <iostream>
#include <string>
#include <cstdlib>
#include <cstdio>
#include <cstring>
#include <vector>
#include <list>

#include <boost/program_options.hpp>
#include <boost/filesystem.hpp>

#include "../jass.hpp"
#include "../i18n.hpp"

namespace
{

const char *version = "0.1";

}

int main(int argc, char *argv[])
{
	// Set the current locale.
	setlocale(LC_ALL, "");
	// Set the text message domain.
	bindtextdomain("jassc", LOCALE_DIR);
	textdomain("jassc");
	
	typedef std::vector<std::string> Strings;
	Strings headers;
	Strings inputFiles;
	boost::filesystem::path outputFile;

	boost::program_options::options_description desc("Allowed options");
	desc.add_options()
	("version,V", _("Shows current version of jassc."))
	("help,h",_("Shows this text."))
	// options
	("header,j", boost::program_options::value<Strings>(&headers), _("Includes JASS headers"))
	// warning options
	("fsyntax-only", _("Check the code for syntax errors, but don't do anything beyond that."))
	
	("i", boost::program_options::value<Strings>(&inputFiles), _("Input files."))
	("o", boost::program_options::value<boost::filesystem::path>(&outputFile), _("Output file."))
	;

	boost::program_options::positional_options_description p;
	p.add("i", -1);

	boost::program_options::variables_map vm;

	try
	{
		boost::program_options::store(boost::program_options::command_line_parser(argc, argv).options(desc).positional(p).run(), vm);
	}
	catch (std::exception &exception)
	{
		std::cerr << boost::format(_("Error while parsing program options: \"%1%\"")) % exception.what() << std::endl;

		return EXIT_FAILURE;
	}

	boost::program_options::notify(vm);

	if (vm.count("version"))
	{
		std::cout <<
		boost::format(_("jassc %1%.")) % version
		<< std::endl <<
		_(
		"Copyright © 2009 Tamino Dauth\n"
		"License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl.html>\n"
		"This is free software: you are free to change and redistribute it.\n"
		"There is NO WARRANTY, to the extent permitted by law."
		) << std::endl;

		return EXIT_SUCCESS;
	}

	if (vm.count("help"))
	{
		std::cout << _("Usage: jassc [options] [input files]") << std::endl << std::endl;
		std::cout << desc << std::endl;
		std::cout << _("\nReport bugs to tamino@cdauth.eu or on https://wc3lib.org") << std::endl;

		return EXIT_SUCCESS;
	}
	
	if (inputFiles.empty()) {
		std::cerr << _("Missing input files.") << std::endl;
		
		return EXIT_FAILURE;
	}
	
	wc3lib::jass::Grammar grammar;
	wc3lib::jass::jass_ast ast;
	bool error = false;
	
	BOOST_FOREACH(Strings::const_reference ref, inputFiles) {
		boost::filesystem::path file(ref);
		
		if (boost::filesystem::is_regular_file(file)) {
			wc3lib::ifstream ifstream(file);
			
			if (ifstream) {
				
				try {
					if (!grammar.parse(ifstream, ast, ref)) {
						error = true;
						
						std::cerr << boost::format(_("Error while parsing file \"%1%\".")) % ref << std::endl;
					}
				} catch (wc3lib::Exception &e) {
					error = true;
					
					std::cerr << boost::format(_("Error while parsing file \"%1%\": %2%")) % ref % e.what() << std::endl;
				}
			}
		} else {
			std::cerr << boost::format(_("File \"%1%\" is no regular file.")) % ref << std::endl;
		}
	}
	
	/*
	 * If an error did occur when parsing the grammar we should not check the AST since
	 * it will be corrupted anyways.
	 * So do only check the AST for types etc. if it had been parsed properly.
	 */
	if (!error) {
		
		wc3lib::jass::Analyzer analyzer(grammar.typeSymbols(), grammar.globalSymbols(), grammar.functionSymbols());
		wc3lib::jass::Analyzer::Reports reports;
		
		analyzer.checkAst(ast, reports);
		
		if (!reports.empty()) {
			BOOST_FOREACH(wc3lib::jass::Analyzer::Reports::const_reference ref, reports) {
				std::cerr << ref.output() << std::endl;
			}
			
			if (vm.count("fsyntax-only")) {
				return EXIT_FAILURE;
			}
		} else {
			std::cout << _("No errors occured.") << std::endl;
			
			if (vm.count("fsyntax-only")) {
				return EXIT_SUCCESS;
			}
		}
	}
	
	return EXIT_SUCCESS;
}
